### ISSUE TITLE
[Bug] running tests in parallel in Docker lead to page crashes

### ISSUE THREAD

**Issue by alexduhanov** [microsoft/playwright#2274]
In the 'page' event [documentation](https://github.com/microsoft/playwright/blob/master/docs/api.md#event-page) there is a note:

>NOTE Use page.waitForLoadState([state[, options]]) to wait until the page gets to a particular state (you should not need it in most cases).

Is there any more information around the possible need for this?

I have three (nearly identical) tests that rarely fail when I run my entire set of tests, specifically in Firefox, and only ever inside of a Docker container.

The tests are quite simple:
1. Navigate to URL
2. Wait for text selector
3. Click on a hyperlink (target=_blank) and wait for 'page' event, return Page object
4. Wait for text selector on returned Page
5. Assert additional text selectors

I'm not sure which step the test fails on, the output simply says "Async callback was not invoked within the 300000ms timeout specified by jest.setTimeout.Timeout". Aside from context and initial page opening and closing, the combined timeouts of all steps in my tests should be 120s, so something seems to be getting stuck.

I'm hoping more information around this note can help me locate the issue.

Thanks!

**Comment by pavelfeldman** [comment]
Your scenario looks solid, there should be no need to wait for the page load state since you already await the selectors you operate on. Could be a bug on our side, any kind of reproduction or additional details would be greatly appreciated.

**Comment by alexduhanov** [comment]
Thanks for the quick reply. Unfortunately I haven't been able to reproduce this deterministically. 

Do you know if there are any issues with running test suites in parallel? Occasionally inside of Docker runs other tests seem to flake out for seemingly no reason with "Target closed" errors. Never had any failed tests when running locally.

**Comment by pavelfeldman** [comment]
Is that all Firefox or other browsers as well? We've seen errors that manifested in a single core environments and never repro in the regular setup. Could you share the details of your host machine and docker image details? We would run our test suite there in a stress test mode.

We don't run our tests in parallel in Docker because Playwright tests are like micro-benchmarks and there was no win due to the CPU being bottleneck. We probably should.

**Comment by alexduhanov** [comment]
The error I mentioned in my original post is so far exclusive to Firefox, but I have different tests that are flaky within Docker in both Firefox and Webkit. Chromium seems be the most stable. There's some pattern to *which* tests fail, but not *why*.

My setup:

Local host machine - MacOS 10.15.4 6 core i7, 32gb ram

CI machine - Oracle Linux 7 running Jenkins on m5.large AWS node

Base docker image - copied from example, minus pwuser setup (received "no sandbox" errors in CI environment). Also copies package.json, babel configuration and runs npm install, for playwright + jest + express + additional dependencies needed to run tests in the container.

Application docker image - points to base image, plus copies compiled application static files and Playwright test files to container at run time.

Application compose file - points to application docker image, sets ipc: host, security_opt to chrome.json file, and binds a volume for access to test report files. (I have tried to remove the volume in the CI environment, but still get the no sandbox error)

10 Jest files, 123 total tests

---

Running docker-compose copies current build/ and tests/ folders to container and runs Jest. Jest globalSetup starts an Express server which runs my application at localhost:8080 inside of the container.

On my local machine, I am able to run all tests in parallel, in all 3 browsers, using Jest describe.each(browserType), consistently with 0 failures.

Inside Docker, I run the tests in parallel, but one browser at a time to save on memory usage.

To summarize:
- Local machine: 0 failures
- Local machine + docker: some flakiness
- CI + docker: some flakiness (first time trying was today)

Please let me know if I can provide any additional information or clarify anything, and let me know if anything about my setup is unusual or might be the source of these issues. 

Thanks so much.

**Comment by aslushnikov** [comment]
@alexduhanov The "target closed" errors you experience seem to be connected to the occasional page crashes. One of the few things that docker constrains is shm size which IIRC defaults to 64MB. Can you try running docker with `docker run --shm-size=1gb` flag?

**Comment by alexduhanov** [comment]
Without setting the flag, the container's /dev/shm is set to 4GB in my CI environment and 2GB on my machine, so not sure if that's the issue.

**Comment by pavelfeldman** [comment]
> Inside Docker, I run the tests in parallel, but one browser at a time to save on memory usage.

How many tests is that? Running tests in parallel means that you are creating as many renderer processes as there are contexts. In WebKit you get extra network processes as well. So yes, you share memory for the browser process, but the rest is pretty much the same. That can put significant pressure to a 2-core machine. If that is too many tests, you can bee OOM-ing or something.

**Comment by alexduhanov** [comment]
I assume that it just runs the test suites in parallel, not the tests within the suites, but I have 10 set suites. Through docker, on my local machine, 5 run at a time and in my CI environment it runs 1 suite at a time. I've added an option to retry failed tests a couple times and in my CI environment it now passes every time (assuming that the one-off failures retry and succeed). Locally, I've still observed some failures, interestingly enough, but I have the local docker container set to 4GB of memory.

**Comment by aslushnikov** [comment]
The discussion here has stalled since we were out of ideas.

@alexduhanov what's the state of the things here? Does it still crash for you?

**Comment by alexduhanov** [comment]
@aslushnikov - I will try running my tests through Docker in parallel again soon and post back. I've been running the tests in serial since I posted the original question and things have generally been fine, though on very rare occasions I still get target closed errors.

**Comment by dgozman** [comment]
Closing since we cannot reproduce the issue, and there is not enough information. If there is some additional information available, please reopen.
